pkgs/development/libraries/blkhash/default.nix

```diff
@@ -0,0 +1,57 @@
+{ lib
+, stdenv
+, fetchFromGitHub
+, fetchFromGitLab
+, meson
+, ninja
+, pkg-config
+, cmake
+, openssl
+, libnbd
+, asciidoc
+, python3Packages
+}:
+let
+  unity = fetchFromGitHub {
+    owner = "ThrowTheSwitch";
+    repo = "Unity";
+    rev = "5204c1bacf37e0c38211d2fbb6d9796e410223f8";
+    sha256 = "i4uQOG/Q86AaQINJoC/65QA/ZEUOXzaVceR2inw32yU=";
+  };
+in
+stdenv.mkDerivation rec {
+  version = "0.6.1";
+  pname = "blkhash";
+
+  src = fetchFromGitLab {
+    owner = "nirs";
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "+7SIZdvkdMYVgeklbqU0MHKc1VJnk9opqhsCmkg6bcI=";
+  };
+
+  postUnpack = ''(
+    cd "$sourceRoot/subprojects"
+    cp -R --no-preserve=mode,ownership ${unity} unity
+  )'';
+
+  dontUseCmakeConfigure = true;
+
+  buildInputs = [
+    meson
+    ninja
+    pkg-config
+    cmake
+    openssl
+    libnbd
+    asciidoc
+  ];
+
+  meta = with lib; {
+    description = "Block based hash optimized for disk images";
+    homepage = "https://gitlab.com/nirs/blkhash";
+    license = licenses.lgpl21Plus;
+    maintainers = with maintainers; [ noisersup ];
+    platforms = platforms.linux;
+  };
```

Comment:
Does this build and install only the library (libblkhash.so) or also the program (blksum) and headers?

Currently the meson.build file does not allow building only some parts. If this is
a problem for Nix, we already have an issue to improve the build system:
https://gitlab.com/nirs/blkhash/-/issues/60

Please add Nix related details to the issue if needed.

Reply:
nixpkgs-review answered my quetion.

Reply:
Oh it seems that I forgot about one dependency...
Fixed by copying the `qemu-nbd` from qemu-utils and checked locally with `pure` flag which clears most of the local environment:

```

Reply:
It looks strange that the blkhash includes the qemu-nbd executable,
instead of requiring installation of the qemu-utils package, but I'm
not sure what are Nix guidlines for dependencies.

```
$ tree /nix/store/lhiks6dp87by27p4q7l8m2plbyfxpqwv-blkhash-0.6.1/bin
/nix/store/lhiks6dp87by27p4q7l8m2plbyfxpqwv-blkhash-0.6.1/bin
├── blksum
└── qemu-nbd
```

Other packages requiring qemu-utils use wrapProgram:
- pkgs/tools/virtualization/alpine-make-vm-image/default.nix
- pkgs/tools/misc/cloud-utils/default.nix
- pkgs/desktops/gnome/apps/gnome-boxes/default.nix
- pkgs/tools/virtualization/multipass/default.nix

alpine-make-vm-image has similar dependency, it uses qemu-img to create images.
Checking how it works shows that it installs also qemu-utils:

```
$ nix-env -i alpine-make-vm-image
...
copying path '/nix/store/nn3hia1xambkv3mas60d641r0zmig0zk-qemu-utils-7.2.0' from 'https://cache.nixos.org'...
```

After the installation, we have:
```
$ ls /nix/store/*-qemu-utils-*/bin/qemu-img
/nix/store/nn3hia1xambkv3mas60d641r0zmig0zk-qemu-utils-7.2.0/bin/qemu-img

$ ls -lha /nix/store/*-alpine-make-vm-image-*/bin
total 24K
dr-xr-xr-x. 2 nsoffer nsoffer  71 Jan  1  1970 .
dr-xr-xr-x. 3 nsoffer nsoffer  17 Jan  1  1970 ..
-r-xr-xr-x. 1 nsoffer nsoffer 784 Jan  1  1970 alpine-make-vm-image
-r-xr-xr-x. 1 nsoffer nsoffer 19K Jan  1  1970 .alpine-make-vm-image-wrapped
```

So it does not include the exexutable. Checking the scripts it does:

```
#! /nix/store/qqa28hmysc23yy081d178jfd9a1yk8aw-bash-5.2-p15/bin/bash -e
export PATH='...:/nix/store/nn3hia1xambkv3mas60d641r0zmig0zk-qemu-utils-7.2.0/bin:...'                                                                                 
exec -a "$0" "/nix/store/lzs1vsgrm77brfn7xg8dqaji61yvbxjc-alpine-make-vm-image-0.9.0/bin/.alpine-make-vm-image-wrapped"  "$@" 
```

The .alpine-make-vm-image-wrapped scripts uses qemu-img.

This looks like the expected result that we would like to have in blkhash, no?

Note: after removing alipine-make-vm-image, qemu-utils is left installed - this looks
like a bug.
```
$ nix-env -e alpine-make-vm-image
uninstalling 'alpine-make-vm-image-0.9.0'
building '/nix/store/qwm7ihnknbvrdijym6l4d5bndsgh446v-user-environment.drv'...

$ nix-collect-garbage 
finding garbage collector roots...
deleting garbage...

$ ls /nix/store/*-qemu-utils-*/bin/qemu-img
/nix/store/nn3hia1xambkv3mas60d641r0zmig0zk-qemu-utils-7.2.0/bin/qemu-img
```

Reply:
Yeah, wrapping binaries seems reasonable here. Fixed it, now qemu-utils are appended to the path for blksum, but not for whole derivation:
```
#! /nix/store/qqa28hmysc23yy081d178jfd9a1yk8aw-bash-5.2-p15/bin/bash -e
export PATH='/nix/store/nn3hia1xambkv3mas60d641r0zmig0zk-qemu-utils-7.2.0/bin'
exec -a "$0" "/nix/store/azwpks48b8wrdci8d2ajdrm7h5smx7xc-blkhash-0.6.1/bin/.blksum-wrapped"  "$@" 
```
> Note: after removing alipine-make-vm-image, qemu-utils is left installed - this looks
like a bug.

Please take a look at `nix-store --query --roots /nix/store/nn3hia1xambkv3mas60d641r0zmig0zk-qemu-utils-7.2.0/bin/qemu-img`, it should tell if path is used somewhere.

I've also added a simple way to disable `libnbd` in derivation, to make it possible to run it on other unix platforms. Let's wait for CI to check if it build correctly!

Reply:
qemu-nbd is need by blksum, not for user. Users can install qemu-utils if needed, so I don't think this is an issue.

However enabling the build for macOS does not work. I suggest to keep it simple and start with Linux. macOS support can be added later.

Reply:
> qemu-nbd is need by blksum, not for user. Users can install qemu-utils if needed, so I don't think this is an issue.

That's why we use `wrapProgram` so `qemu-nbd` is only accessible for blksum, am I missing something?

Reply:
No, I think what we have now is fine.